### PR TITLE
Added OutputFormat setting date_time_output_format

### PR DESCRIPTION
**Pull request by kitaisreal**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

https://github.com/ClickHouse/ClickHouse/issues/10862

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added setting date_time_output_format

### CODE REVIEW

File: src/DataTypes/DataTypeDateTime.cpp [hunk 1]
```diff
@@ -59,9 +59,20 @@ String DataTypeDateTime::doGetName() const
...
+    switch (settings.date_time_output_format)
+    {
+        case FormatSettings::DateTimeOutputFormat::Simple:
+            writeDateTimeText(assert_cast<const ColumnType &>(column).getData()[row_num], ostr, time_zone);
+            return;
+        case FormatSettings::DateTimeOutputFormat::UnixTimestamp:
+            writeDateTimeTextUnixTimestamp(assert_cast<const ColumnType &>(column).getData()[row_num], ostr, time_zone);
+            return;
+        case FormatSettings::DateTimeOutputFormat::ISO:
+            /// TODO: Implement
```
Comment (kitaisreal) [inline src/DataTypes/DataTypeDateTime.cpp:73]: @alexey-milovidov Could you please provide some references to search in existing code base if we already have implementation for printing date in ISO format? Thanks.

Comment (alexey-milovidov) [reply src/DataTypes/DataTypeDateTime.cpp:73]: We don't have it, but it's very easy to implement.

It should print DateTime in the following form: `2020-10-11T23:42:13Z`
You can implement it similar to writeDateTimeText, but you must use UTC timezone.

File: src/IO/WriteHelpers.h [hunk 2]
```diff
@@ -798,6 +798,28 @@ inline void writeDateTimeTextRFC1123(time_t datetime, WriteBuffer & buf, const D
     buf.write(" GMT", 4);
 }
 
+inline void writeText(const String & x, WriteBuffer & buf) { writeEscapedString(x, buf); }
+
+inline void writeDateTimeTextUnixTimestamp(time_t datetime, WriteBuffer & buf, const DateLUTImpl & date_lut = DateLUT::instance())
```
Comment (alexey-milovidov) [inline src/IO/WriteHelpers.h:803]: You don't need this function.

unix timestamp is the `time_t` itself.
It is independent of time zone.

File: src/IO/WriteHelpers.h [hunk 3]
```diff
@@ -798,6 +798,28 @@ inline void writeDateTimeTextRFC1123(time_t datetime, WriteBuffer & buf, const D
...
+{
+    const auto & values = date_lut.getValues(datetime);
+    auto local_date_time_unix_timestamp = static_cast<time_t>(LocalDateTime(values.year, values.month, 
+        values.day_of_month, date_lut.toHour(datetime), date_lut.toMinute(datetime), date_lut.toSecond(datetime)));
+    writeText(std::to_string(local_date_time_unix_timestamp), buf);
+}
+
+inline void
+writeDateTimeTextUnixTimestamp(DateTime64 datetime64, UInt32 scale, WriteBuffer & buf, const DateLUTImpl & date_lut = DateLUT::instance())
```
Comment (alexey-milovidov) [inline src/IO/WriteHelpers.h:812]: Here you should write unix_timestamp.fractional_part.

File: src/IO/WriteHelpers.h [hunk 4]
```diff
@@ -828,8 +850,6 @@ template <typename T>
 inline std::enable_if_t<std::is_floating_point_v<T>, void>
 writeText(const T & x, WriteBuffer & buf) { writeFloatText(x, buf); }
 
-inline void writeText(const String & x, WriteBuffer & buf) { writeEscapedString(x, buf); }
```
Comment (alexey-milovidov) [inline src/IO/WriteHelpers.h:831]: ?

File: src/DataTypes/DataTypeDateTime64.cpp [hunk 5]
```diff
@@ -57,9 +57,20 @@ std::string DataTypeDateTime64::doGetName() const
...
+void DataTypeDateTime64::serializeText(const IColumn & column, size_t row_num, WriteBuffer & ostr, const FormatSettings & settings) const
 {
-    writeDateTimeText(assert_cast<const ColumnType &>(column).getData()[row_num], scale, ostr, time_zone);
+    switch (settings.date_time_output_format)
+    {
+        case FormatSettings::DateTimeOutputFormat::Simple:
+            writeDateTimeText(assert_cast<const ColumnType &>(column).getData()[row_num], scale, ostr, time_zone);
+            return;
+        case FormatSettings::DateTimeOutputFormat::UnixTimestamp:
+            writeDateTimeTextUnixTimestamp(assert_cast<const ColumnType &>(column).getData()[row_num], scale, ostr, time_zone);
```
Comment (alexey-milovidov) [inline src/DataTypes/DataTypeDateTime64.cpp:68]: `writeIntText` will output what you need.
No need to implement `writeDateTimeTextUnixTimestamp` function.

File: src/IO/WriteHelpers.h (hunk 2 above)
Comment (kitaisreal) [reply src/IO/WriteHelpers.h]: Fixed. Thanks.

File: src/DataTypes/DataTypeDateTime.cpp (hunk 1 above)
Comment (kitaisreal) [reply src/DataTypes/DataTypeDateTime.cpp]: Added function in WriteHelpers.h

File: src/IO/WriteHelpers.h (hunk 4 above)
Comment (kitaisreal) [reply src/IO/WriteHelpers.h]: I moved it at the top of file. Fixed.

File: src/DataTypes/DataTypeDateTime64.cpp (hunk 5 above)
Comment (kitaisreal) [reply src/DataTypes/DataTypeDateTime64.cpp]: Fixed. Thanks.

File: src/IO/WriteHelpers.h (hunk 3 above)
Comment (kitaisreal) [reply src/IO/WriteHelpers.h]: Currently toUnixTimestamp function if applied to DateTime64 column does not write fractional_part in addition to unix_timestamp. Should we write that ?
```
CREATE TABLE test_date_time_64(timestamp DateTime64) ENGINE=Log;
INSERT INTO test_date_time_64 VALUES (1546300800000);
SELECT toUnixTimestamp(timestamp) FROM test_date_time_64;


┌─toUnixTimestamp(timestamp)─┐
│                 1546300800 │
└────────────────────────────┘
```

Comment (alexey-milovidov) [reply src/IO/WriteHelpers.h]: It's in question how this function should work. Now it does not return fractional part, it's Ok (it returns unix timestamp, indeed).

File: src/IO/WriteHelpers.h [hunk 6]
```diff
@@ -798,6 +798,36 @@ inline void writeDateTimeTextRFC1123(time_t datetime, WriteBuffer & buf, const D
...
+{
+    static constexpr UInt32 MaxScale = DecimalUtils::maxPrecision<DateTime64>();
+    scale = scale > MaxScale ? MaxScale : scale;
+
+    auto c = DecimalUtils::split(datetime64, scale);
+    const auto & values = utc_time_zone.getValues(c.whole);
+    writeDateTimeText<'-', ':', 'T'>(
+        LocalDateTime(values.year, values.month, values.day_of_month,
+            utc_time_zone.toHour(c.whole), utc_time_zone.toMinute(c.whole), utc_time_zone.toSecond(c.whole)), buf);
+    buf.write('Z');
```
Comment (alexey-milovidov) [inline src/IO/WriteHelpers.h:820]: But for DateTime64 we need to write fractional part as well:
`2019-08-20T10:18:56.012345Z`

File: src/IO/WriteHelpers.h [hunk 7]
```diff
@@ -798,6 +798,36 @@ inline void writeDateTimeTextRFC1123(time_t datetime, WriteBuffer & buf, const D
...
+}
+
+inline void writeDateTimeUnixTimestamp(DateTime64 datetime64, UInt32 scale, WriteBuffer & buf)
+{
+    static constexpr UInt32 MaxScale = DecimalUtils::maxPrecision<DateTime64>();
+    scale = scale > MaxScale ? MaxScale : scale;
+
+    auto c = DecimalUtils::split(datetime64, scale);
+    writeIntText(c.whole, buf);
```
Comment (alexey-milovidov) [inline src/IO/WriteHelpers.h:829]: Let's write fractional part.

**Comment by alexey-milovidov** [comment]
Let's also add a test, see `tests/queries/0_stateless`

File: src/IO/WriteHelpers.h (hunk 6 above)
Comment (kitaisreal) [reply src/IO/WriteHelpers.h:815]: Fixed. Thanks.

File: src/IO/WriteHelpers.h (hunk 7 above)
Comment (kitaisreal) [reply src/IO/WriteHelpers.h:824]: Updated.

**Comment by kitaisreal** [comment]
@alexey-milovidov added tests.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Now everything looks alright :)

**Pull request merged**